File: example/web/src/hooks.ts
import type { Transport } from '@sveltejs/kit/hooks';

/**
 * Money is the browser's half of the domain type. It is a class, not an
 * interface, because the point of the `transport` hook is that an instance
 * arrives with its behaviour: a page calls `price.format()`.
 *
 * The Go half is `businesslogic.Money`, declared in `src/hooks.go`.
 */
export class Money {
	readonly cents: number;

	constructor(cents: number) {
		this.cents = cents;
	}

	format(): string {
		const sign = this.cents < 0 ? '-' : '';
		const cents = Math.abs(this.cents);
		return `${sign}$${Math.floor(cents / 100)}.${String(cents % 100).padStart(2, '0')}`;
	}
}

/**
 * transport is kit's universal hook for custom types. Each key is the tag the
 * value travels under, and it has to be spelled the same in `src/hooks.go`:
 * Go serializes a Money as `["Money", …]`, and this is what reads it back.
 *
 * `encode` returns `false` for anything that is not a Money, which is how kit
 * asks "is this yours?". What it returns for a match is what the Go decoder
 * receives when a Money travels the other way, in a remote function's argument.
 */
export const transport: Transport = {
	Money: {
		encode: (value) => value instanceof Money && { cents: value.cents },
		decode: (data: { cents: number }) => new Money(data.cents)
	}
};

File: example/web/src/hooks.go
// Package hooks is the app's universal hooks, the Go half of what
// `src/hooks.ts` declares beside it.
//
// Kit's `transport` is a universal hook because both sides of the wire need it:
// the browser needs `decode` to rebuild a class instance, and the server needs
// `encode` to say what that instance is made of. The two halves are one
// declaration, so they live next to each other.
package hooks

import (
	"github.com/tylergannon/skgo"
	"github.com/tylergannon/skgo/example/businesslogic"
)

// Money crosses the wire as a custom type. Without this a price arrives in the
// browser as a plain `{ cents: 2000 }` and `price.format()` throws.
var _ = skgo.Transported[businesslogic.Money]("Money")
